src/nccl_ofi_net.c

```diff
@@ -1404,84 +1409,75 @@ static ncclResult_t ofi_connect(int dev, void *handle, void **sendComm)
 	req->dev = sComm->dev;
 	req->direction = NCCL_OFI_SEND;
 
+	conn_info = (struct connection_info*)calloc(1, sizeof(*conn_info));
+	if (OFI_UNLIKELY(conn_info == NULL)) {
+		NCCL_OFI_WARN("Couldn't allocate connection_info for dev %d", dev);
+		ret = ncclSystemError;
+		goto error;
+	}
+	conn_info->ep_namelen = sizeof(conn_info->ep_name);
+
 	/* Get local EP address to transfer in the connect message */
 	ret = fi_getname(&(nccl_ofi_component[dev]->ep->fid),
-			 (void *)&local_ep_addr,
-			 &namelen);
+			 conn_info->ep_name,
+			 &(conn_info->ep_namelen));
 	if (ret != 0) {
 		NCCL_OFI_WARN("Call to fi_getname() failed with RC: %d, ERROR: %s",
 			      ret, fi_strerror(-ret));
 		ret = ncclSystemError;
 		goto error;
 	}
 
+	connect_to_self = (memcmp(conn_info->ep_name, remote_ep_name, conn_info->ep_namelen) == 0);
+
+	if (connect_to_self) {
+		conn_info->req = req;
+		/* for future freeing conn_info buffer */
+		conn_info->memory_buffer = conn_info;
+	}
+
 	/* Send "connect" message to remote EP */
-	/* do not wait for completion in case of sending to self */
-	if (memcmp(local_ep_addr, remote_ep_addr, namelen) == 0) {
-		static char buffer_to_self[MAX_EP_ADDR];
-		NCCL_OFI_TRACE(NCCL_NET, "Connect to self");
-		if (request_to_self != NULL) {
-			NCCL_OFI_WARN("Another connect to self request already in progress");
-			ret = ncclSystemError;
-			goto error;
-		}
-		memcpy(buffer_to_self, local_ep_addr, MAX_EP_ADDR);
-		rc = fi_tsend(sComm->local_ep, buffer_to_self,
-		              MAX_EP_ADDR, NULL, sComm->remote_ep,
-		              sComm->tag | ~max_tag, &req->ctx);
-		if (rc == 0) {
-			NCCL_OFI_TRACE(NCCL_NET, "Connect to self request completed");
-		}
+	do {
+		/*
+		 * TODO: replace it with API of FI_INJECT type when most of
+		 * providers can support it, so that need for completion check
+		 * below can be lifted.
+		 */
+		rc = fi_tsend(sComm->local_ep, conn_info,
+			      sizeof(*conn_info), NULL, sComm->remote_ep,
+			      sComm->tag | ~max_tag, &req->ctx);
+		if (rc == 0)
+			break;
 		else if (rc == -FI_EAGAIN) {
-			NCCL_OFI_TRACE(NCCL_NET, "Connect to self request will be completed once receiver is ready");
-			/* postpone request completion, setting req to NULL prevents freeing it at the end */
-			request_to_self = req;
-			req = NULL;
+			/*
+			 * Process completions so that you have enough
+			 * resources for sending connect message
+			 */
+			ret = nccl_ofi_progress(nccl_ofi_component[dev]);
+			if (OFI_UNLIKELY(ret != 0))
+				goto error;
 		}
 		else {
-			NCCL_OFI_WARN("Unable to send connect message to self for dev %d. RC: %zd, ERROR: %s",
-			              dev, rc, fi_strerror(-rc));
+			NCCL_OFI_WARN("Unable to send connect message for dev %d. RC: %zd, ERROR: %s",
+				     dev, rc, fi_strerror(-rc));
 			ret = ncclSystemError;
 			goto error;
 		}
+	} while (true);
+
+	/* Ensure the message is sent. */
+	/* Skip this step for sending to self */
+	if (connect_to_self) {
```

Comment:
This generalizes the handling of self connections for all providers. Basically, no matter whether the provider has synchronous or asynchronous behavior, we would always handle completions for self requests in the accept phase. This requires us to do some internal testing to ensure we do not break other supported providers. We will update the PR once we have the results.

Reply:
Looking forward your results, thanks.